hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.client.heartbeat;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.util.HeartbeatUtils;
+import org.apache.hudi.common.util.ValidationUtils;
+import org.apache.hudi.exception.HoodieException;
+import org.apache.hudi.exception.HoodieHeartbeatException;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.stream.Collectors;
+
+/**
+ * This class creates heartbeat for hudi client. This heartbeat is used to ascertain whether the running job is or not.
```

Comment:
@n3nash spent sometime thinking about this. the tradeoffs are pretty clear in the sense, this helps us reclaim the storage used by failed writes much sooner than say deleting the files, during archival process (which by itself is not a guarantee actually).

If we want to implement some schema to pre-emptively take away compute resources used by a pending job or storage space, then I don't see a way to do this without heartbeats of some kind. 

We could just keep this simpler in this PR, if you like and just clean up once during archival. (again, the case of a long deletion job running for say 12 hrs, comes to mind. and That needs some kind of heartbeat mechanism, so that even archival does not clean up). 

Orthogonal question : for such long running concurrent writes, should n't they just abort early if they detect more commits have happened since it began?  i.e if it figures out after 1 hr of runtime, that some writes have changed the same filegroups, then why should it even run? its going to fail anyway

Reply:
will review this class again, pending this discussion here.

Reply:
High level, no other way to guard against these cases. Even if we check for long running concurrent writes and resolve every N heartbeats, backfill jobs may not have anything overlapping and may even OOM at the end since folks don't know how to tune large spark jobs.